### PR TITLE
add dart chapter_array_and_linkedlist 

### PR DESCRIPTION
**Pull request by JeffersonHuang**
If this PR is related to coding or code translation, please read the [contribution guideline](https://github.com/krahets/hello-algo/issues/15#issue-1464069591), fill out the checklist, and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

File: codes/dart/chapter_array_and_linkedlist/my_list.dart [hunk 1]
```diff
@@ -0,0 +1,135 @@
+/**
+ * File: my_list.dart
+ * Created Time: 2023-02-05
+ * Author: Jefferson (JeffersonHuang77@gmail.com)
+ */
+class MyList {
```
Comment (krahets) [inline codes/dart/chapter_array_and_linkedlist/my_list.dart:6]: Add an empty line above.

File: codes/dart/chapter_array_and_linkedlist/my_list.dart [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  /// [_nums]  数组（存储列表元素）
+  late List<int> _nums;
+
+  /// [_capacity]列表容量
+  int _capacity = 10;
```
Comment (krahets) [inline codes/dart/chapter_array_and_linkedlist/my_list.dart:11]: Can we use 4 spaces for indentation in Dart? If so, please use 4 spaces to keep the consistency with other languages.

File: codes/dart/chapter_array_and_linkedlist/array.dart [hunk 3]
```diff
@@ -0,0 +1,110 @@
+/**
+ * File: array.dart
+ * Created Time: 2023-01-20
+ * Author: Jefferson (JeffersonHuang77@gmail.com)
+ */
+
+import 'dart:math';
```
Comment (krahets) [inline codes/dart/chapter_array_and_linkedlist/array.dart:7]: Could we create a `utils/include.dart` and import all the common libs in this file?

File: codes/dart/chapter_array_and_linkedlist/array.dart [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+class Array {
+
+  /// 随机返回一个 数组元素
```
Comment (krahets) [inline codes/dart/chapter_array_and_linkedlist/array.dart:11]: Please use `/* ... */` to comment on the class and function headers.

```dart
/* 标题注释，用于标注函数、类、测试样例等 */

// 内容注释，用于详解代码

/**
 * 多行
 * 注释
 */
```

File: codes/dart/chapter_array_and_linkedlist/list.dart [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import 'dart:ffi';
+
+/**
```
Comment (krahets) [inline codes/dart/chapter_array_and_linkedlist/list.dart:3]: File header should be placed at the top of the file.

File: codes/dart/chapter_array_and_linkedlist/my_list.dart [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    for (var j = index; j < _size - 1; j++) {
+      _nums[j] = _nums[j + 1];
+    }
+    // 更新元素数量
+    _size--;
+    // 返回被删除元素
+    return num;
+  }
+
+  void extendCapacity() {
```
Comment (krahets) [inline codes/dart/chapter_array_and_linkedlist/my_list.dart:80]: function comment missing

File: codes/dart/utils/PrintUtil.dart [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import '../chapter_array_and_linkedlist/array.dart';
```
Comment (krahets) [inline codes/dart/utils/PrintUtil.dart:1]: Is it necessary to include array.dart?

File: codes/dart/utils/PrintUtil.dart [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import 'ListNode.dart';
+
+/**
```
Comment (krahets) [inline codes/dart/utils/PrintUtil.dart:4]: File header on the top

File: codes/dart/chapter_array_and_linkedlist/array.dart [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+    // 将原数组中的所有元素复制到新数组
+    for (var i = 0; i < nums.length; i++) {
+      res[i] = nums[i];
+    }
+    // 返回扩展后的新数组
+    return res;
+  }
+
+  /// 在数组的索引[index] 处插入元素[num]
```
Comment (krahets) [inline codes/dart/chapter_array_and_linkedlist/array.dart:33]: Recommend `在数组的索引 index 处插入元素 num` to keep the consistency among multi-languages. We might use find&replace to update comments for the codes of different languages.

**Review by krahets — COMMENTED** [review]
Thanks, @JeffersonHuang! Please address the comments.

File: codes/dart/chapter_array_and_linkedlist/my_list.dart (hunk 2 above)
Comment (JeffersonHuang) [reply codes/dart/chapter_array_and_linkedlist/my_list.dart]: Yes, 4 spaces can be used, but this is against the dart style specification. Are you sure you need to replace?

File: codes/dart/chapter_array_and_linkedlist/array.dart (hunk 3 above)
Comment (JeffersonHuang) [reply codes/dart/chapter_array_and_linkedlist/array.dart:7]: No, this way will throw an error.

File: codes/dart/chapter_array_and_linkedlist/my_list.dart (hunk 2 above)
Comment (krahets) [reply codes/dart/chapter_array_and_linkedlist/my_list.dart]: No, let us keep 2 spaces.

**Review by krahets — APPROVED** [review]
Thanks @JeffersonHuang! Looking forward to more PRs related to Dart translation.

**Pull request merged**